### DISCUSSION TITLE
How cadence do keep execution history order?

### DISCUSSION THREAD

**Discussion by zedongh** [uber/cadence#4684] [Q&A] [answered]
For sequence workflow, I could understand the EventID is incr 1 by 1, and ResetWorkflowExecution opeartion is clear.

But for parallel workflow (using workflow.Go function), how cadence keep the EventID always be same evenly in history replay?
In such case, reset workflow seems be confusing, because the history event not a simple sequence structure.

 If I want to reset just single one branch of parallel, but other branch history event  happend after the event id point, what will happen?

**Comment by longquanzheng** [comment, marked as answer]
The trick is within the SDK and has nothing to do with server. 
SDK will make sure all the goroutines( they are called coroutine in the Go SDK) are executed in a deterministic order. 
And that's exactly why it's required to use `workflow.Go` because native `go func...` will lead to non-deterministic error. 

For more details, see this dispatcherImpl  https://github.com/uber-go/cadence-client/blob/26294594c6dd6b74e8d626a929968cace8e3148a/internal/internal_workflow.go#L157 
Which provides ExecuteUntilAllBlocked 
https://github.com/uber-go/cadence-client/blob/26294594c6dd6b74e8d626a929968cace8e3148a/internal/internal_workflow.go#L901

**Reply by zedongh** [reply]
Thanks!
